File: broker.go/internal/archive/manager.go
package archive

import (
	"context"
	"fmt"
	"log/slog"
	"regexp"
	"sync"
	"time"

	"monstermq.io/edge/internal/config"
	"monstermq.io/edge/internal/stores"
	storememory "monstermq.io/edge/internal/stores/memory"
	storemongo "monstermq.io/edge/internal/stores/mongodb"
	storepg "monstermq.io/edge/internal/stores/postgres"
	storesqlite "monstermq.io/edge/internal/stores/sqlite"
)

// Manager owns the active set of archive groups and dispatches every published
// message to all matching groups. It supports SQLite, PostgreSQL and MongoDB
// per-group last-value and archive stores.
type Manager struct {
	cfg      *config.Config
	storage  *stores.Storage
	logger   *slog.Logger
	sqliteDB *storesqlite.DB
	pgDB     *storepg.DB
	mongoDB  *storemongo.DB

	mu          sync.RWMutex
	groups      map[string]*Group
	deployError map[string]string // last deploy error per group name (for the dashboard)
}

func NewManager(cfg *config.Config, storage *stores.Storage,
	sqliteDB *storesqlite.DB, pgDB *storepg.DB, mongoDB *storemongo.DB,
	logger *slog.Logger) *Manager {
	return &Manager{
		cfg: cfg, storage: storage, logger: logger,
		sqliteDB: sqliteDB, pgDB: pgDB, mongoDB: mongoDB,
		groups:      map[string]*Group{},
		deployError: map[string]string{},
	}
}

// DeployError returns the last error encountered while trying to start the
// named group, or empty string if the group is healthy.
func (m *Manager) DeployError(name string) string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.deployError[name]
}

// Load creates archive groups from the persistent ArchiveConfigStore. If no
// "Default" group exists yet, one is provisioned with the broker's primary
// backend (so a Postgres-backed broker gets a Postgres-backed Default group).
func (m *Manager) Load(ctx context.Context) error {
	configs, err := m.storage.ArchiveConfig.GetAll(ctx)
	if err != nil {
		return err
	}
	hasDefault := false
	for _, c := range configs {
		if c.Name == "Default" {
			hasDefault = true
			break
		}
	}
	if !hasDefault {
		def := stores.ArchiveGroupConfig{
			Name:          "Default",
			Enabled:       true,
			TopicFilters:  []string{"#"},
			LastValType:   m.defaultLastValType(),
			ArchiveType:   m.defaultArchiveType(),
			PayloadFormat: stores.PayloadDefault,
		}
		if err := m.storage.ArchiveConfig.Save(ctx, def); err != nil {
			return err
		}
		configs = append(configs, def)
	}
	for _, c := range configs {
		if err := m.startGroup(ctx, c); err != nil {
			m.recordDeployError(c.Name, err)
			m.logger.Error("archive group start failed", "name", c.Name, "err", err)
		}
	}
	return nil
}

func (m *Manager) recordDeployError(name string, err error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if err == nil {
		delete(m.deployError, name)
		return
	}
	m.deployError[name] = err.Error()
}

func (m *Manager) defaultLastValType() stores.MessageStoreType {
	switch m.cfg.DefaultStoreType {
	case config.StorePostgres:
		return stores.MessageStorePostgres
	case config.StoreMongoDB:
		return stores.MessageStoreMongoDB
	}
	return stores.MessageStoreSQLite
}

func (m *Manager) defaultArchiveType() stores.MessageArchiveType {
	switch m.cfg.DefaultStoreType {
	case config.StorePostgres:
		return stores.ArchivePostgres
	case config.StoreMongoDB:
		return stores.ArchiveMongoDB
	}
	return stores.ArchiveSQLite
}

func (m *Manager) startGroup(ctx context.Context, c stores.ArchiveGroupConfig) error {
	if !c.Enabled {
		return nil
	}
	lastVal, err := m.buildLastValStore(ctx, c)
	if err != nil {
		return err
	}
	arch, err := m.buildArchiveStore(ctx, c)
	if err != nil {
		return err
	}
	flushInterval := time.Duration(m.cfg.Archive.FlushIntervalMs) * time.Millisecond
	g := NewGroup(c, lastVal, arch,
		m.cfg.Archive.BufferSize, m.cfg.Archive.MaxBatchSize, flushInterval,
		m.logger)
	g.Start()
	m.mu.Lock()
	m.groups[c.Name] = g
	delete(m.deployError, c.Name)
	m.mu.Unlock()
	m.logger.Info("archive group started",
		"name", c.Name, "filters", c.TopicFilters,
		"lastValType", c.LastValType, "archiveType", c.ArchiveType)
	return nil
}

// Per-group store and archive names match the Kotlin/JVM broker:
//   last-value store →  "<groupName>Lastval"   (SQL table: <groupname>lastval)
//   message archive  →  "<groupName>Archive"   (SQL table: <groupname>archive)
// so the same physical database can be opened by either implementation.

// LastValName / ArchiveName are exported so tests and the dashboard can derive
// the same names production uses. Both are deterministic from the group name.
func LastValName(group string) string { return group + "Lastval" }
func ArchiveName(group string) string { return group + "Archive" }

// validGroupNamePattern restricts group names to characters that are safe to
// embed in a SQL identifier. Group names flow into CREATE TABLE / DROP TABLE
// statements via fmt.Sprintf, so anything outside this set could be SQL
// injection.
var validGroupNamePattern = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9_]{0,62}$`)

// ValidateGroupName returns an error if name is not a safe SQL identifier
// fragment. Called from the GraphQL create/update resolvers.
func ValidateGroupName(name string) error {
	if !validGroupNamePattern.MatchString(name) {
		return fmt.Errorf("invalid group name %q: must match [A-Za-z][A-Za-z0-9_]{0,62}", name)
	}
	return nil
}

func (m *Manager) buildLastValStore(ctx context.Context, c stores.ArchiveGroupConfig) (stores.MessageStore, error) {
	name := LastValName(c.Name)
	switch c.LastValType {
	case stores.MessageStoreMemory:
		return storememory.NewMessageStore(name), nil
	case stores.MessageStoreSQLite:
		if m.sqliteDB == nil {
			return nil, fmt.Errorf("group %s: lastValType=SQLITE but no SQLite DB configured", c.Name)
		}
		s := storesqlite.NewMessageStore(name, m.sqliteDB)
		if err := s.EnsureTable(ctx); err != nil {
			return nil, fmt.Errorf("ensure %s on sqlite: %w", name, err)
		}
		return s, nil
	case stores.MessageStorePostgres:
		if m.pgDB == nil {
			return nil, fmt.Errorf("group %s: lastValType=POSTGRES but no Postgres DB configured", c.Name)
		}
		s := storepg.NewMessageStore(name, m.pgDB)
		if err := s.EnsureTable(ctx); err != nil {
			return nil, fmt.Errorf("ensure %s on postgres: %w", name, err)
		}
		return s, nil
	case stores.MessageStoreMongoDB:
		if m.mongoDB == nil {
			return nil, fmt.Errorf("group %s: lastValType=MONGODB but no MongoDB configured", c.Name)
		}
		s := storemongo.NewMessageStore(name, m.mongoDB)
		if err := s.EnsureTable(ctx); err != nil {
			return nil, fmt.Errorf("ensure %s on mongo: %w", name, err)
		}
		return s, nil
	case stores.MessageStoreNone, "":
		return nil, nil
	}
	return nil, fmt.Errorf("group %s: unsupported lastValType %q", c.Name, c.LastValType)
}

func (m *Manager) buildArchiveStore(ctx context.Context, c stores.ArchiveGroupConfig) (stores.MessageArchive, error) {
	name := ArchiveName(c.Name)
	switch c.ArchiveType {
	case stores.ArchiveSQLite:
		if m.sqliteDB == nil {
			return nil, fmt.Errorf("group %s: archiveType=SQLITE but no SQLite DB configured", c.Name)
		}
		a := storesqlite.NewMessageArchive(name, m.sqliteDB, c.PayloadFormat)
		if err := a.EnsureTable(ctx); err != nil {
			return nil, fmt.Errorf("ensure %s on sqlite: %w", name, err)
		}
		return a, nil
	case stores.ArchivePostgres:
		if m.pgDB == nil {
			return nil, fmt.Errorf("group %s: archiveType=POSTGRES but no Postgres DB configured", c.Name)
		}
		a := storepg.NewMessageArchive(name, m.pgDB, c.PayloadFormat)
		if err := a.EnsureTable(ctx); err != nil {
			return nil, fmt.Errorf("ensure %s on postgres: %w", name, err)
		}
		return a, nil
	case stores.ArchiveMongoDB:
		if m.mongoDB == nil {
			return nil, fmt.Errorf("group %s: archiveType=MONGODB but no MongoDB configured", c.Name)
		}
		a := storemongo.NewMessageArchive(name, m.mongoDB, c.PayloadFormat)
		if err := a.EnsureTable(ctx); err != nil {
			return nil, fmt.Errorf("ensure %s on mongo: %w", name, err)
		}
		return a, nil
	case stores.ArchiveNone, "":
		return nil, nil
	}
	return nil, fmt.Errorf("group %s: unsupported archiveType %q", c.Name, c.ArchiveType)
}

// Reload re-reads the archive config table and reconciles the running groups
// with the persisted state: starts new ones, restarts changed ones, stops
// removed ones. Call this after any GraphQL mutation that touched an archive
// group config.
func (m *Manager) Reload(ctx context.Context) error {
	configs, err := m.storage.ArchiveConfig.GetAll(ctx)
	if err != nil {
		return err
	}
	wanted := map[string]stores.ArchiveGroupConfig{}
	for _, c := range configs {
		if c.Enabled {
			wanted[c.Name] = c
		}
	}

	m.mu.Lock()
	current := m.groups
	m.mu.Unlock()

	// Stop groups that are gone or now disabled or whose config changed.
	for name, g := range current {
		w, keep := wanted[name]
		if !keep || !configsEqual(g.Config(), w) {
			g.Stop()
			m.mu.Lock()
			delete(m.groups, name)
			m.mu.Unlock()
		}
	}
	// Start groups that should be running but aren't.
	for name, c := range wanted {
		m.mu.RLock()
		_, running := m.groups[name]
		m.mu.RUnlock()
		if running {
			continue
		}
		if err := m.startGroup(ctx, c); err != nil {
			m.recordDeployError(name, err)
			m.logger.Error("archive group reload failed", "name", name, "err", err)
		}
	}
	return nil
}

func configsEqual(a, b stores.ArchiveGroupConfig) bool {
	if a.Name != b.Name || a.Enabled != b.Enabled || a.RetainedOnly != b.RetainedOnly {
		return false
	}
	if a.LastValType != b.LastValType || a.ArchiveType != b.ArchiveType {
		return false
	}
	if a.PayloadFormat != b.PayloadFormat {
		return false
	}
	if a.LastValRetention != b.LastValRetention || a.ArchiveRetention != b.ArchiveRetention {
		return false
	}
	if len(a.TopicFilters) != len(b.TopicFilters) {
		return false
	}
	for i := range a.TopicFilters {
		if a.TopicFilters[i] != b.TopicFilters[i] {
			return false
		}
	}
	return true
}

// Dispatch routes msg to every matching group.
func (m *Manager) Dispatch(msg stores.BrokerMessage) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	for _, g := range m.groups {
		if g.Matches(msg.TopicName, msg.IsRetain) {
			g.Submit(msg)
		}
	}
}

func (m *Manager) Stop() {
	m.mu.Lock()
	defer m.mu.Unlock()
	for _, g := range m.groups {
		g.Stop()
	}
	m.groups = map[string]*Group{}
}

// Snapshot returns a stable view of the current groups for resolvers.
func (m *Manager) Snapshot() []*Group {
	m.mu.RLock()
	defer m.mu.RUnlock()
	out := make([]*Group, 0, len(m.groups))
	for _, g := range m.groups {
		out = append(out, g)
	}
	return out
}

File: broker.rs/src/archive/manager.rs
use super::group::Group;
use crate::config::{Config, StoreType};
use crate::stores::memory::MemoryMessageStore;
use crate::stores::sqlite::Db as SqliteDb;
use crate::stores::sqlite::{SqliteMessageArchive, SqliteMessageStore};
use crate::stores::storage::Storage;
use crate::stores::traits::{MessageArchive, MessageStore};
use crate::stores::types::{ArchiveGroupConfig, BrokerMessage, MessageArchiveType, MessageStoreType, PayloadFormat};
use once_cell::sync::Lazy;
use parking_lot::RwLock;
use regex::Regex;
use std::collections::HashMap;
use std::sync::Arc;

pub fn last_val_name(group: &str) -> String { format!("{}Lastval", group) }
pub fn archive_name(group: &str) -> String { format!("{}Archive", group) }

static VALID_GROUP_NAME: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[A-Za-z][A-Za-z0-9_]{0,62}$").unwrap());

/// Validate a group name — names flow into CREATE TABLE / DROP TABLE via
/// string interpolation so anything outside this set is rejected.
pub fn validate_group_name(name: &str) -> anyhow::Result<()> {
    if !VALID_GROUP_NAME.is_match(name) {
        anyhow::bail!("invalid group name {:?}: must match [A-Za-z][A-Za-z0-9_]{{0,62}}", name);
    }
    Ok(())
}

/// Manager owns the active set of archive groups and dispatches every
/// published message to all matching groups.
pub struct Manager {
    cfg: Config,
    storage: Storage,
    sqlite_db: Option<SqliteDb>,
    groups: RwLock<HashMap<String, Arc<Group>>>,
    deploy_errors: RwLock<HashMap<String, String>>,
}

impl Manager {
    pub fn new(cfg: Config, storage: Storage, sqlite_db: Option<SqliteDb>) -> Arc<Self> {
        Arc::new(Self {
            cfg, storage, sqlite_db,
            groups: RwLock::new(HashMap::new()),
            deploy_errors: RwLock::new(HashMap::new()),
        })
    }

    pub fn deploy_error(&self, name: &str) -> Option<String> {
        self.deploy_errors.read().get(name).cloned()
    }

    fn default_last_val_type(&self) -> MessageStoreType {
        match self.cfg.default_store_type {
            StoreType::Postgres => MessageStoreType::Postgres,
            StoreType::Mongodb => MessageStoreType::Mongodb,
            _ => MessageStoreType::Sqlite,
        }
    }

    fn default_archive_type(&self) -> MessageArchiveType {
        match self.cfg.default_store_type {
            StoreType::Postgres => MessageArchiveType::Postgres,
            StoreType::Mongodb => MessageArchiveType::Mongodb,
            _ => MessageArchiveType::Sqlite,
        }
    }

    pub async fn load(&self) -> anyhow::Result<()> {
        let mut configs = self.storage.archive_config.get_all().await?;
        if !configs.iter().any(|c| c.name == "Default") {
            let def = ArchiveGroupConfig {
                name: "Default".to_string(),
                enabled: true,
                topic_filters: vec!["#".to_string()],
                last_val_type: self.default_last_val_type(),
                archive_type: self.default_archive_type(),
                payload_format: PayloadFormat::Default,
                ..Default::default()
            };
            self.storage.archive_config.save(&def).await?;
            configs.push(def);
        }
        for c in configs {
            if let Err(e) = self.start_group(c.clone()).await {
                self.record_error(&c.name, Some(e.to_string()));
                tracing::error!(name = c.name, err = %e, "archive group start failed");
            }
        }
        Ok(())
    }

    fn record_error(&self, name: &str, err: Option<String>) {
        let mut w = self.deploy_errors.write();
        match err {
            Some(s) => { w.insert(name.to_string(), s); }
            None => { w.remove(name); }
        }
    }

    async fn build_last_val_store(&self, c: &ArchiveGroupConfig) -> anyhow::Result<Option<Arc<dyn MessageStore>>> {
        let name = last_val_name(&c.name);
        match c.last_val_type {
            MessageStoreType::Memory => Ok(Some(Arc::new(MemoryMessageStore::new(name)) as Arc<dyn MessageStore>)),
            MessageStoreType::Sqlite => {
                let db = self.sqlite_db.clone().ok_or_else(|| anyhow::anyhow!("group {}: lastValType=SQLITE but no SQLite DB configured", c.name))?;
                let s = Arc::new(SqliteMessageStore::new(name, db));
                s.ensure_table().await?;
                Ok(Some(s as Arc<dyn MessageStore>))
            }
            MessageStoreType::None => Ok(None),
            other => anyhow::bail!("group {}: lastValType {:?} not implemented in broker.rs yet", c.name, other),
        }
    }

    async fn build_archive_store(&self, c: &ArchiveGroupConfig) -> anyhow::Result<Option<Arc<dyn MessageArchive>>> {
        let name = archive_name(&c.name);
        match c.archive_type {
            MessageArchiveType::Sqlite => {
                let db = self.sqlite_db.clone().ok_or_else(|| anyhow::anyhow!("group {}: archiveType=SQLITE but no SQLite DB configured", c.name))?;
                let a = Arc::new(SqliteMessageArchive::new(name, db, c.payload_format));
                a.ensure_table().await?;
                Ok(Some(a as Arc<dyn MessageArchive>))
            }
            MessageArchiveType::None => Ok(None),
            other => anyhow::bail!("group {}: archiveType {:?} not implemented in broker.rs yet", c.name, other),
        }
    }

    async fn start_group(&self, c: ArchiveGroupConfig) -> anyhow::Result<()> {
        if !c.enabled { return Ok(()); }
        let lv = self.build_last_val_store(&c).await?;
        let ar = self.build_archive_store(&c).await?;
        let g = Group::new(c.clone(), lv, ar);
        g.start();
        self.groups.write().insert(c.name.clone(), g);
        self.record_error(&c.name, None);
        tracing::info!(name = c.name, filters = ?c.topic_filters, last_val = c.last_val_type.as_str(), archive = c.archive_type.as_str(), "archive group started");
        Ok(())
    }

    pub async fn reload(&self) -> anyhow::Result<()> {
        let configs = self.storage.archive_config.get_all().await?;
        let wanted: HashMap<String, ArchiveGroupConfig> = configs.into_iter()
            .filter(|c| c.enabled)
            .map(|c| (c.name.clone(), c))
            .collect();

        let to_stop: Vec<(String, Arc<Group>)> = {
            let r = self.groups.read();
            r.iter().filter_map(|(k, v)| {
                let stop = match wanted.get(k) {
                    None => true,
                    Some(w) => !configs_equal(v.config(), w),
                };
                if stop { Some((k.clone(), v.clone())) } else { None }
            }).collect()
        };
        for (name, g) in to_stop {
            g.stop().await;
            self.groups.write().remove(&name);
        }
        for (name, c) in wanted {
            if self.groups.read().contains_key(&name) { continue; }
            if let Err(e) = self.start_group(c).await {
                self.record_error(&name, Some(e.to_string()));
                tracing::error!(name, err = %e, "archive group reload failed");
            }
        }
        Ok(())
    }

    /// Deliver `msg` to every matching live group.
    pub fn dispatch(&self, msg: BrokerMessage) {
        let r = self.groups.read();
        for g in r.values() {
            if g.matches(&msg.topic_name, msg.is_retain) {
                g.submit(msg.clone());
            }
        }
    }

    pub async fn stop(&self) {
        let groups: Vec<Arc<Group>> = self.groups.read().values().cloned().collect();
        for g in groups { g.stop().await; }
        self.groups.write().clear();
    }

    pub fn snapshot(&self) -> Vec<Arc<Group>> {
        self.groups.read().values().cloned().collect()
    }
}

fn configs_equal(a: &ArchiveGroupConfig, b: &ArchiveGroupConfig) -> bool {
    a.name == b.name
        && a.enabled == b.enabled
        && a.retained_only == b.retained_only
        && a.last_val_type == b.last_val_type
        && a.archive_type == b.archive_type
        && a.payload_format == b.payload_format
        && a.last_val_retention == b.last_val_retention
        && a.archive_retention == b.archive_retention
        && a.topic_filters == b.topic_filters
}
